### PR TITLE
Support more use cases for 'out' parameter

### PR DESCRIPTION
**Pull request by antonwolfy**
The PR is about to improve handling of `out` parameter in supported `dpnp` functions.
It's required to consider use cases when
- `out` refers to the same array as any of input ones
- `out` with strides
- `out` where dtype is different from expected one

In both cases extra memory is allocated and it's need to copy the result back from temporary array into `out` array.

- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Review by vlad-perevezentsev — APPROVED** [review]
Good job!

File: dpnp/dpnp_utils/dpnp_algo_utils.pyx [hunk 1]
```diff
@@ -740,6 +745,14 @@ cdef class dpnp_descriptor:
     def is_scalar(self):
         return self.dpnp_descriptor_is_scalar
 
+    @property
+    def is_tempored(self):
```
Comment (oleksandr-pavlyk) [inline dpnp/dpnp_utils/dpnp_algo_utils.pyx:749]:
```suggestion
    def is_temporary(self):
```

File: dpnp/dpnp_utils/dpnp_algo_utils.pyx [hunk 2]
```diff
@@ -784,6 +797,13 @@ cdef class dpnp_descriptor:
             "expected either dpctl.tensor.usm_ndarray or dpnp.dpnp_array.dpnp_array, got {}"
             "".format(type(self.origin_pyobj)))
 
+    def get_result_desc(self):
+        if self.is_tempored:
```
Comment (oleksandr-pavlyk) [inline dpnp/dpnp_utils/dpnp_algo_utils.pyx:801]:
```suggestion
        if self.is_temporary:
```

**Review by vlad-perevezentsev — APPROVED** [review]
LGTM!

**Pull request merged**